Topology aware dynamic provisioning design

msau42 opened the pull request:
This design is for 1.11 work to support dynamic provisioning.

Update volume topology design doc with:
* Defining terms used to describe topology
* Dynamic provisioning new APIs
* In-tree and external provisioning interface changes
* Scheduler implementation details
* Examples for local, zonal, and multi-zonal volumes

@kubernetes/sig-storage-pr-reviews 

/hold

msau42 commented:
/assign @jsafrane @saad-ali @thockin @lichuqiang

lichuqiang commented on contributors/design-proposals/storage/volume-topology-scheduling.md:
```diff
@@ -266,41 +304,329 @@ const (
+}
+
+type StorageClassCapacity struct {
+    TopologyKey string
```
Still I'm concerned about the Topology key in StorageClassCapacity, if we place it in the status field, and leave it to provisioners to specify, then for edge-case that:
 
 When scheduler find a storage with StorageClassCapacity empty, it is hard to tell whether the storage is indeed without capacity limit, or it's just of a state that capacity has not been initialized by provisioners yet.

 If we move the Topology key out of status field, and leave it to admin, it is thought to be fulfilled on StorageClass creation.
 And the edge-case above could be avoid.

msau42 replied:
Then it will open up a new mode for failure, where administrator specifies the wrong topologyKey and the provisioner values won't match with it.

If a pod gets scheduled before provisioners have reported their capacity, then:
* Provisioning could fail.  In that case we should retry scheduling
* Provisioning could succeed.  Then we will not have marked the PVC with provisioned capacity information, so our calculations for used capacity will be incorrect.  Is there a way we can handle this?  Ie, have a StorageClass modify handler, when StorageClassCapacity is initialized, then initialize the capacity cache?

lichuqiang replied:
Oh, I would prefer the scheduler could recognize the StorageClass at the first beginning, but not manage things by trying Provisioning. That makes things complex and may have effect on performance.

If you are worried that admin may specify a wrong key, then a simple flag to imply whether the StorageClass is of  capacity limit may also work.
I think the admin should at least know the StorageClass he creates is of  capacity limit or not.

msau42 replied:
How will the scheduler know that all the provisioners have initialized and reported capacity? (in the case of local volume provisioners)

msau42 replied:
Nm, in that case, once the topologyKey is specified, then capacity is zero until otherwise updated.

lichuqiang replied:
> How will the scheduler know that all the provisioners have initialized and reported capacity? (in the case of local volume provisioners)

I think it doesn't need to know, what the scheduler cares:
1. Whether the StorageClass is of capacity limit
2. If so, whether it's capacity could satisfy the requirement.

No matter it's capacity smaller than requirement, or it's in state of uninitialized, it's both in state of unavailable for scheduler.

msau42 replied:
I discussed with @saad-ali on this.  Our consensus is that it's better for topologyKey to be reported by the provisioner.  Depending on the user to specify the topologyKey, or even a flag, opens up the possibility that they could incorrectly set it.  And they don't have any choice in what to set as the value.  For local storage, it must always be set to "kubernetes.io/hostname", there is no other value that will work.

In the worst case, there will be a period of time during cluster initialization where the provisioners may not have reported capacity yet.  The consequences are not fatal.  Provisioning could fail until the provisioners are up and running (which is expected behavior in general).  Provisioning could succeed, and in that case we'll have to resync the capacity cache later.  We could try to mitigate this scenario, by recommending that provisioners do not process any provisioning requests until they have fully initialized (including reporting capacity).

msau42 commented:
TODO: Add section about predicate ecache invalidation and active pod queues

redbaron commented:
Failure domain for storage in general case can be different from failure domain for compute. Lets say in bare-metal setup, there can be 2 SAN storage, each available to half of compute nodes, while compute nodes themselves are split into failure domains per-rack.

There is a hack for Local volumes in current kubernetes version, where the fact that storage failure domain is per node is special cased. Do  think  suggestion to add first class storage-specific failure domains is something which belongs to this proposal?

msau42 commented:
/assign @verult

verult commented on contributors/design-proposals/storage/volume-topology-scheduling.md:
```diff
@@ -266,41 +304,329 @@ const (
+Some volume types have capacity limitations per topology domain. For example,
+local volumes have limited capacity per node. Provisioners for these volume
+types can report these capacity limitations through a new StorageClass status
+field.
```
I've been warned before against having constantly changing metrics (like resource usage) be placed as an API status, and one argument is that it could easily overwhelm the API server. Maybe the capacity information should be plumbed to the scheduler through existing resource usage monitoring pipelines.

Might be good to get a review from sig-instrumentation?

msau42 replied:
The capacity reported by the provisioner should be total capacity, not available capacity.  I don't anticipate total capacity changing very frequently.

The challenge with existing resource usage tracking is that it is done on a per node level, whereas volume resource limits may not necessarily be at the node boundary.

verult replied:
There are plans to eventually plumb PVC usage stats through Metrics Server, as part of the work to get these stats exposed to kubectl. I believe the scheduler also talks to the Metrics Server.

Currently there's a limitation that PVC usage stats can only be measured when the underlying volume is mounted in a pod, but maybe we can expand this feature and add the capability of having volume plugins report these stats as well.

@jingxu97

msau42 replied:
I believe for PVC usage stats, it's still tied to a Pod that is scheduled to a specific node.  That way you can hit all the kubelet metrics endpoints.  For this design to handle getting capacity from metrics, we would need to collect all the metrics from a central entity, and not per node to handle volumes with topologies spanning more than a single node.

I am unaware of the scheduler talking to the metrics server.  My understanding is it uses resource capacity reported in the Node object.  @bsalamat could you clarify?

But besides the initial StorageClass creation, I don't anticipate reported capacity to be updated very frequently.

lichuqiang replied:
Can we find the details of the feature somewhere?

msau42 replied:
Here, we have also have the challenge that external provisioners need to report their capacity somewhere, whether on an API object (like proposed), or to some central server.

lichuqiang replied:
Yeah, as I'm also focusing on design of local storage provisioning, I wonder what the way to talk to the metrics server looks like.

I guess users are more likely to look forward to things they are familiar

msau42 replied:
I discussed a bit with @verult, and we agreed that going the metrics server approach would require a lot of extra plumbing that doesn't exist today.  So we're still leaning towards the current approach of reporting total (not used) capacity in the StorageClass and calculating and caching used from provisioned PVCs.

lichuqiang commented on contributors/design-proposals/storage/volume-topology-scheduling.md:
```diff
@@ -266,41 +304,329 @@ const (
+kube-scheduler, kube-controller-manager, and all kubelets.
+
+StorageClass.AllowedTopology and StorageClassCapacity fields will be controlled
+by the DynamicProvisioningScheduling feature gate, and must be configured in the
```
Seem the two are not completely independent, maybe we should control both binding and provisioning with a same Feature Gate?

If we enable `VolumeScheduling` and disable `DynamicProvisioningScheduling`, then what's the expected behavior in pv controller? for PVCs of mode `WaitForFirstConsumer`, skip binding but exec provisioning directly?

msau42 replied:
Challenge is static local PVs is beta, and dynamic provisioning will be alpha, so we can't have one feature gate control both.  I see VolumeScheduling as a superset, so if that is disabled, then the delayed dynamic provisioning is also disabled.  Once delayed dynamic provisioning goes to beta, we can combine the two feature gates.

msau42 commented on contributors/design-proposals/storage/volume-topology-scheduling.md:
```diff
@@ -266,41 +304,329 @@ const (
+}
+
+type VolumeProvisioningTopology struct {
+    Required map[string]VolumeTopologyValues
```
@saad-ali found a use case that this cannot represent:
An administrator wants to restrict provisioning to either "zone A, rack 1" or "zone B, rack 1".

To fix this, we could go back to using a restricted node selector (only support the IN operator), or follow the CSI [proposed](https://github.com/container-storage-interface/spec/pull/188#issuecomment-366425374) representation which sort of does the reverse of the node selector: OR, then AND.

A restricted NodeSelector would be more familiar to Kubernetes users, so I'm leaning towards that approach.   As long as we restrict the operators to just IN, then it can be translated internally into the CSI representation.

msau42 replied:
I'm flip flopping again.  A node selector specifies a subset off of a set of labels on the node object.  But in this case, we are not selecting from a set of labels on an existing object.  In this case, we want to limit the topology of a PV object that will be created in the future.  So I think following the CSI representation as a list of maps makes more sense, rather then specifying a selector on nothing.

lichuqiang replied:
+1

verult replied:
+1. Another semantic difference is: VolumeTopology is used to select **a set of nodes** at which a PV is available, whereas NodeSelector is used to select **a single node**.

liggitt replied:
> NodeSelector is used to select a single node

that's a little surprising... selectors typically select a set of things

verult replied:
@liggitt you are right, I was using a wrong frame of thinking, so I take that back.

Here's another point, similar to what @msau42  said, that helped me understand the difference: ultimately the scheduler needs a set of nodes to choose from; a NodeSelector **directly** specifies that set, whereas VolumeTopology **indirectly** specifies it by providing a super set which is then refined further by the volume plugin. Please correct me if this is the wrong understanding.

lichuqiang commented on contributors/design-proposals/storage/volume-topology-scheduling.md:
```diff
@@ -404,26 +753,30 @@ AssumePVCs(pod *v1.Pod, node *v1.Node) (pvcBindingRequired bool, err error)
     2. If the prebind fails, revert the cache updates.
 2. For in-tree and external dynamic provisioning:
-    1. Set `annStorageProvisioner` on the PVC.
+    1. Set `annScheduledNode` on the PVC.
```
annSelectedNode

kmova commented on contributors/design-proposals/storage/volume-topology-scheduling.md:
````diff
@@ -266,41 +304,329 @@ const (
+    persistentVolumeClaim:
+      claimName: my-pvc
+```
+5. Scheduler picks a node that can satisfy the Pod and passes it to the
````
@msau42 - Can the scheduler pass the "suggested" nodes where it can be scheduled? What if the node that is picked by the scheduler can't be used by the dynamic provisioner to create a PV. 

In the next step (6), how about the dynamic provisioner create a PV and set nodeAffinity (list of nodes, racks, zones) where it can be accessed and the Pod get scheduled on one of those nodes. This could tie into the way, static PVs provisioned and associated with nodeAffinity works.

lichuqiang replied:
> What if the node that is picked by the scheduler can't be used by the dynamic provisioner to create a PV.

I think the topology and capacity limitation of the provisioner should have been specified in the Storageclass.

> In the next step (6), how about the dynamic provisioner create a PV and set nodeAffinity (list of nodes, racks, zones) where it can be accessed and the Pod get scheduled on one of those nodes.

This way, there is still a chance that the provisioned pv is conflict to the scheduled node.
I think the main purpose of the feature is to reconcile pod scheduling with volume binding/provisioning,

msau42 replied:
The general scheduler design requires that we have all the scheduler predicates and priorities we need so that the scheduler can make the right decision.  If there is some timing issue which causes the scheduler's view to be out of date, provisioning needs to fail, and the scheduler has to retry.  Eventually, the scheduler's view of the world has to be corrected and it will make the right choice.

If provisioning doesn't fail and it ends up provisioning in a topology domain that the scheduler didn't recommend, in the best case, the scheduler can still find a node that can satisfy the PV.  In the worst case, the pod will fail scheduling and stay pending.

Step 6 will work exactly like as you mentioned.  The PV provisioned must also specify the nodeAffinity, and on pod restarts, it will follow the same bound PVC flow.

kmova replied:
Thanks @msau42 @laishiquan. Having the topology information to the StorageClass and using that as a scheduler predicate might just work.

msau42 replied:
Can you clarify the use case you are trying to support?  This proposal adds 2 ways to influence scheduling for dynamic provisioning:

1. Admin can restrict topology via AllowedTopologies
2. Provisioner can restrict topology via StorageClassCapacity

msau42 replied:
Do the nodes where the StoragePool exist map to the volume io storage pods?  Also, can a node serve multiple storage pools?

I'm trying to see if you could use the storage class capacity reporting instead of having the admin set allowedTopologies, because that will open up the possibility that the admin will configure it incorrectly.  But the capacity reporting scheme may not work if you can have a node serve multiple storage pools because you cannot have multiple label values for a single key.

msau42 replied:
One more question. Is it a requirement that the application pod is on the same node where the storagepool is in?

msau42 replied:
Hm IIUC, you will have one storageclass per storage pool.  In that case, then capacity reporting won't help, unless the provisioner does something like report some fake capacity for each node in the storagepool (and by default 0 capacity for the rest)

kmova replied:
A node can server multiple StoragePools depending on the available storage and other resources on the node. 

Since the communication is like this: application pod -> volume io controller pod (stateless) -> storagepool (storage io pod), it is not a hard requirement to have the application pod on the same node as the storagepool, but it is a preferred way. 

@msau42 it is a good point though about the possibility that the admin may configure incorrectly, to circumvent this - we could use an operator to handle the updates to the allowedToplogies. 

Is the the allowedTopologies an editable parameter on the StorageClass object?

msau42 replied:
We could make it mutable. I don't see a reason not to.  Do you see a use case where the admin would want to configure some policy that is different from the operator though?  Maybe they would still want to allow more nodes outside of the storagepool because they are node resource constrained.  

It sounds like for your use case, topology preferences would be better than requirements.

msau42 replied:
Originally, the allowedTopologies was intended to be set by the user, hence, having it in the "spec".  And then capacity is intended to be reported by the system, therefore it's in the "status".  The use case you bring up doesn't fit quite nicely into this model. I need to think about it more

msau42 replied:
@kmova I pondered some more about your use case and I'm not actually sure you need to influence topology at all during provisioning.  Let me know if this makes any sense:

You have one storageclass per storagepool.  A user requests a specific storageclass ie a specific storagepool.  Since it's not a requirement that pods land on the same node where a storagepool spans, you should always have provisioning succeed even if it results in suboptimal placement.  After provisioning, you can specify your topology preferences in the PV.nodeaffinity (note, preferences are not supported today and would need to be added), and the scheduler can use the preferences to make the final node choice.

orainxiong commented on contributors/design-proposals/storage/volume-topology-scheduling.md:
````diff
@@ -383,19 +732,19 @@ PrioritizeUnboundPVCs(pod *v1.Pod, filteredNodes HostPriorityList) (rankedNodes
-AssumePVCs(pod *v1.Pod, node *v1.Node) (pvcBindingRequired bool, err error)
+AssumeVolumes(pod *v1.pod, node *v1.node) (pvcbindingrequired bool, err error)
 ```
 1. If all the Pod’s PVCs are bound, return false.
````
I think the rough logic of AssumePodVolumes would be : 

- Could returns true if all volumes are fully bound
- Could returns false if any volume binding operation needs to be done afterward, and then the pod will be enqueued into _BindQueue_

orainxiong commented:
I have modified _volume-topology-scheduling.png_ a little.
[https://www.processon.com/view/link/59e6fc25e4b040dc8507bfd1](url)
Hope it helps.

msau42 commented:
Updated AllowedTopologies to align with how CSI specifies topology, and should have addressed all the review comments

msau42 commented on contributors/design-proposals/storage/volume-topology-scheduling.md:
```diff
@@ -266,41 +304,368 @@ const (
+If no StorageClassCapacity is specified, then there are no capacity
+limitations.
+
+Capacity reporting is limited to a single topology key. `TopologyKey`
```
I think it doesn't have to be restricted to a single key if the Capacities map key is a VolumeTopology type.  Not sure if that is expressible in the API though, I haven't seen any non-string map keys in the API.

If we can't do the above, then Capacities could be an array, but then lookups by the scheduler will be inefficient.  We need to figure out how important capacity reporting is for multi-dimensional storage systems.  The examples I can think of are only hypothetical at this point.

lichuqiang replied:
This would indeed bring inconvenience to provisioners and the scheduler,
I would suggest we do not introduce this before we can see practical use case of multi-dimensional  capacity reporting

msau42 replied:
We can support multi-keys but still have efficient lookups by flattening the values into a single string separated by a delimiter.  Will update the spec shortly.

msau42 replied:
Also there is a limitation that a single storage pool should not be shared by multiple storage classes, otherwise you will have more capacity reported than is actually available.

verult commented on contributors/design-proposals/storage/volume-topology-scheduling.md:
````diff
@@ -328,30 +693,54 @@ avoid these error conditions are to:
+
+In-tree provisioners:
+```
+Provision(selectedNode *string, allowedTopologies *storagev1.VolumeProvisioningTopology) (*v1.PersistentVolume, error)
````
What if the entire Node spec is passed instead of just the node name? A topology-aware volume plugin has to fetch the node object anyway in order to get the node's topology info.

Alternatively - should there be a dedicated NodeTopology field inside Node spec? Then here we could just pass that field, and volume plugins encode their topology information in this field rather than labels.

A related issue: https://github.com/kubernetes/kubernetes/issues/61906 (TL;DR if we had this field, it could be passed to applications through downwardAPI, making it easy for applications to obtain topology info).

msau42 replied:
Having a dedicated topology field in the node object could be useful, but I think it has broader implications than just for storage, so would need to be considered more holistically.

Passing a Node object sounds ok to me. It means that the PV controller also needs permissions to get Node objects.

lichuqiang replied:
@msau42 As we discussed before, we need a way to calculate the capacity used by released PVs(i.e. provisioned PVs whose claims got deleted then)

so here we'll also need to pass `AnnProvisionedTopology`, and require the developers of provisioners to add the annotation onto PVs.

msau42 replied:
done

msau42 commented on contributors/design-proposals/storage/volume-topology-scheduling.md:
```diff
@@ -266,41 +304,368 @@ const (
+  constraints that restrict it to 'zone1' and 'zone3'.  If 'zone1'
+  fails, the Pod cannot be scheduled to the intended failover zone.
+
+Kubernetes will leave validation and enforcement of the AllowedTopologies content up
```
Labels are user defined and can be anything.  And each volume plugin may have a different idea of what topology labels it needs to operate on.

verult replied:
But in the end these are directly compared against Node labels for scheduling purposes, so they should at least satisfy the Kubernetes label format. I'd imagine the reasoning is similar to why NodeSelector and NodeAffinity are validated against the label format as well.

jsafrane commented on contributors/design-proposals/storage/volume-topology-scheduling.md:
```diff
@@ -266,41 +304,368 @@ const (
+  requiredToplogies:
+  - matchNodeLabels:
+      failure-domain.beta.kubernetes.io/zone: us-central1-a
+  - matchNodeLabels:
```
It took me a while to understand why there are two `matchNodeLabels` entries above. It could be useful to explicitly note somewhere that Kubernetes computes OR between matchNodeLabels and AND among individual labels inside.

msau42 replied:
Updated in API comments

jsafrane commented on contributors/design-proposals/storage/volume-topology-scheduling.md:
```diff
@@ -266,41 +304,368 @@ const (
+restrictions specified under plugin-specific parameters to the new AllowedTopologies
+field.
+
+The AllowedTopologies will also be provided to provisioners as a new field, detailed in
```
It would be nice to have a sentence or two why node selector was not used in StorageClass.

msau42 replied:
added

jsafrane commented on contributors/design-proposals/storage/volume-topology-scheduling.md:
```diff
@@ -266,41 +304,368 @@ const (
+* StorageClass object size can be big for large clusters. Assuming that
+  each node entry in the Capacities map could take up around 100 bytes
+  (size of node name + resource.Quantity), a 1000 node cluster that uses
+  dynamic provisioning of local volumes could add 100Ki to the StorageClass
```
Just brainstorming: what if we move the StorageClassCapacity into its own object? For local storage there would be one StorageClass as it is today + every node would have its own StorageClassCapacity somehow linked to the StorageClass. Individual local storage provisioners would not conflict when updating their own StorageClassCapacity objects. On the other hand, multiple small objects would probably load API server more than one big one.

jsafrane replied:
Ack, let's leave it in StorageClass.

jsafrane commented on contributors/design-proposals/storage/volume-topology-scheduling.md:
```diff
@@ -266,41 +304,368 @@ const (
+  scenario, the provisioner should fail provisioning until it has reported its
+  capacity and not process any PVCs that have not been updated with the chosen
+  topologyKey value.
+
```
Another downsides:

- When the storage is shared among several Kubernetes clusters *or* with something outside Kubernetes, Kubernetes cannot reliably compute amount of used storage. It can compute only the storage the particular cluster consumes (i.e. PVs), but it won't see PVs of other clusters and/or users outside of Kubernetes. It is recommended that Kubernetes gets dedicated storage or storage quota so it can reliably compute amount of used/remaining storage just by counting PVs.

- The storage may be shared by several StorageClasses. Typically, Gluster allows admins to choose nr. of replicas of PVs in StorageClass. There may be two StorageClasses that use the same Gluster servers, however each of them creates PVs with different nr. of replicas. Kubernetes treats StorageClasses as independent on each other, however in this case PVs in one StorageClass must be counted to used space in a different StorageClass.

- Capacity does not report granularity of the storage. E.g. there may be 1 GiB of available space on Gluster in a particular datacenter and scheduler can schedule one pod with two 0.5 GiB PVCs there. Gluster allocates storage in 1GiB chunks, so it can't accommodate both PVCs and the pod becomes unschedulable until someone frees up some space in the datacenter.
  - This is IMO rare corner case when a storage is almost full.

jsafrane replied:
> I propose to instead probe the storage provider if needed: "Could you create a volume of size X and type Y?" This could be a "dry run" variant of the volume create api.

Scheduler is under watch for its performance. Currently, it does not call any external API. Calling number of probe/dry-run methods would slow everything down. We need a quick and mostly reliable check where to start provisioning first, hoping it succeeds in vast majority of cases.

msau42-tmp replied:
Another problem with doing a call out is that the "provisioning check" and "actual provisioning" steps are not synchronous.  So the following two scenarios would not be accurate:

* Pod requests multiple PVCs.  You would need to query the storage system in a way so that the 1st query would temporarily subtract from the total when doing the 2nd query.
* Scheduler does the actual binding operations asynchronously, so you could have a query for the next Pod before the provisioning for the 1st Pod is triggered.  This is similar to the 1st scenario, just for different Pods.

That's why the scheduler today doesn't do callouts to get the Node resources, and it tracks the used capacity internally.  So if we want to do something similar for storage, then I think we should follow this similar model of storing total capacity in an object.  I agree that there are many internal details that influence what is actually provisionable, so if we had such an API, it would be up to the provisioner to report its capacity wrt to the user's request.  However, the limitation of the current capacity API is that the storage pool is contiguous, which doesn't sound to be the case with gluster (discrete disks).

jsafrane commented on contributors/design-proposals/storage/volume-topology-scheduling.md:
````diff
@@ -266,41 +304,368 @@ const (
+      node2: 200Gi
+```
+
+TODO: should this be in the local volume spec instead?
````
How this can be in local PV spec? We shouldn't put capacity of a node into all local PVs that are the node.

jsafrane commented on contributors/design-proposals/storage/volume-topology-scheduling.md:
```diff
@@ -328,30 +693,54 @@ avoid these error conditions are to:
-`annStorageProvisioner` annotation in the PVC.
+`annSelectedNode` annotation in the PVC. If provisioning fails, the PV
+controller can signal back to the scheduler to retry dynamic provisioning by
+removing the `annSelectedNode` annotation.
```
"removing the `annSelectedNode` annotation" is not described below. Who removes it and when? And can scheduler pick the same node again or will it ensure that it picks a different node this time?

msau42 commented on contributors/design-proposals/storage/volume-topology-scheduling.md:
```diff
@@ -194,12 +244,11 @@ and region labels.  This will handle newly created PV objects.
 Existing PV objects will have to be upgraded to use the new NodeAffinity field.
 This does not have to occur instantaneously, and can be updated within the
-deprecation period.
+deprecation period.  This can be done through a new PV update controller that
```
Added API update section

msau42 commented on contributors/design-proposals/storage/volume-topology-scheduling.md:
```diff
@@ -266,41 +304,502 @@ const (
+
+## API Upgrades
+
+### Zonal PV Upgrade
```
Add an example for multi-zonal PV upgrade because the zone label value format is special.

msau42 commented on contributors/design-proposals/storage/volume-topology-scheduling.md:
```diff
@@ -141,6 +185,9 @@ spec:
 
 #### Zonal Volume
+In this example, the volume can only be accessed from nodes that have the
+label key `failure-domain.beta.kubernetes.io/zone' and label value
```
done thanks

msau42 commented on contributors/design-proposals/storage/volume-topology-scheduling.md:
```diff
@@ -266,41 +304,565 @@ const (
+...
+status:
+  capacity:
+    topologyKey: zone,rack
```
Keys

lichuqiang commented on contributors/design-proposals/storage/volume-topology-scheduling.md:
```diff
@@ -266,41 +304,565 @@ const (
+type StorageClass struct {
+    ...
+
+    Status *StorageClassStatus
```
generally speaking, status field should not be a pointer?

thockin commented on contributors/design-proposals/storage/volume-topology-scheduling.md:
```diff
@@ -266,41 +304,565 @@ const (
+limitations.  A missing map entry for a topology domain will be treated
+as zero capacity.
+
+A scheduler predicate will use the reported capacity to determine if dynamic
```
Also it will scale badly for large clusters.

thockin replied:
Also it is insufficient to express whether a given volume can actually fit (e.g. could be fragmented).

msau42 replied:
This is similar to how a node today reports total CPU/memory.  It doesn't actually report used vs available, instead the scheduler keeps track of used based on all the Pods that have been scheduled to that node.  This way you avoid the latency issue of waiting for reported values to updated after they are allocated.

So we're trying to follow a similar concept here, except the challenge with volumes is that their topology is not always scoped to a single node, so it doesn't make sense to report as part of node capacity.  The reported capacity would only change if you change the capacity of the storage pool, but not when you allocate/free volumes from it.  I expect the former will happen less frequently than the latter, so the scaling limits should be similar to how we process cpu/memory capacity.

But this does have some limitations to be accurate, like you mentioned:
* The storage pool has to be dedicated to this StorageClass and cannot be shared with other StorageClasses
* Provisioning from the storage pool cannot be prone to fragmentation

msau42 replied:
Here's an alternative idea.  Considering that only local volumes plan to use the capacity reporting api at the moment, and there are some restrictions on how you deploy the storage pools to ensure that this capacity handling is actually accurate, we could try going with a CRD + scheduler extender approach.

The downsides are:
- Deployment is more complicated
- We can't include this into Kubernetes core.  The provisioner is already out of tree, maybe this not an issue.
- The predicate will become best-effort, due to the delay of seeing the PVC provisioning API update in the extender.  The fact that the scheduler round robins on equal priority nodes could help mitigate this, so that the same node is not chosen again for the next Pod.

Advantages are:
- Since it's a CRD specific to local volumes, the API could be simpler and more flexible because we don't need to make it generic to all volume topologies
- You potentially could support scenarios like storage pools being shared across multiple StorageClasses, since it doesn't need to be strictly tied to a StorageClass anymore
- You can encode special behavior/requirements of your specific storage system into the extender logic.
- Keeps StorageClass lean and we don't have to regret not putting a Spec/Status in the beginning

@lichuqiang @jsafrane @thockin wdyt?

lichuqiang commented on contributors/design-proposals/storage/volume-topology-scheduling.md:
```diff
@@ -266,41 +304,565 @@ const (
+}
+
+type StorageClassCapacity struct {
+    // Minimum size of a provisioned volume
```
As @jsafrane said, "some clusters allocates storage in 1GiB chunks", thus, if a user requested 1.5GiB in PVC, we are expected to reject it, even if we have enough capacity left.
Agree we might need to consider whether the capacity field is a good place to specify it.

lichuqiang replied:
Well, I misunderstood its meaning here, and thought we need to reject it if a user requested a volume under the minimum size.

jsafrane replied:
IMO it should be admin decision what "rounding" is acceptable and what's too much. Someone finds using 50GB for 10GB of data just fine, others find it waste of space.

CSI introduced `required_bytes` and `limit_bytes` and I like it, however we don't have capability to express that in PVC and in storage class, where both user and admin might define what's acceptable.

msau42 replied:
PVC already has the fields to support request and limit, we just don't use limit currently.  I'm still not quite sure about the use case of putting it in the storageclass.  I believe for the other resources, admins can define a limit range to restrict at admission.

jsafrane commented on contributors/design-proposals/storage/volume-topology-scheduling.md:
```diff
@@ -266,41 +304,565 @@ const (
 provisioning is not used (in the case of local storage).
-* We have to maintain two different paths for volume binding.
+* We have to maintain two different code paths for volume binding.
 * We will be depending on the storage admin to correctly configure the
```
I don't have any special use case for very slow-to-bind volumes, it would just break existing behavior. Users (and automation) are used to their PVCs being bound relatively quickly and they will be confused.

Someone on SIG-storage said that in some storage backends provisioning can take hours and it's well planned operation, so users want to create PVCs and let them provision before they start a pod. @kangarlou, was that you?

kangarlou replied:
@jsafrane No, it wasn't me, but here is my two cents: 
Today for dynamic provisioners, binding happens as soon as the provisioning is done. For the NFS and iSCSI volumes that we manage on NetApp storage, provisioning happens very quickly. To keep it fast, our provisioner doesn't format the iSCSI volumes and relies on kubelet to perform this functionality during attach. This is not a concern for us, but I can see cases where provisioning takes a long time for some plugins/provisioners (e.g., formatting large drives at the time of volume creation, provisioning from snapshots). Having said that, I don't think we should maintain two code paths. As far as external automation is concerned, because provisioning can take arbitrarily long time (in the worst case, the provisioning can fail), any external tooling should inspect the PVC status and not rely on the immediate binding behavior. I agree with @jsafrane that some users prefer to provision their PVCs before they start the pods, but for provisioners that take a long time, I don't think the proposed changes affect the user experience much as such users see provisioning and binding as one step, not as two separate steps.

kangarlou commented on contributors/design-proposals/storage/volume-topology-scheduling.md:
```diff
@@ -328,30 +890,62 @@ avoid these error conditions are to:
 unbound PVCs with VolumBindingWaitForFirstConsumer and no prebound PVs
 to let it come through the scheduler path.
 
 Dynamic provisioning will also be skipped if
```
Dynamic provisioning will be skipped or delayed until `annSelectedNode` annotation is set? Does this mean dynamic provisioners should look for this annotation for the storage classes with `VolumeBindingMode` set to `VolumeBindingWaitForFirstConsumer` before they can act on the PVCs of that type?

msau42 commented:
I squashed the PR into 3 commits:
* Design up the most recent review iteration
* Addressed feedback from the most recent review, changing allowedTopologies to use a new TopologySelector field instead of an array of maps
* Separated out the capacity reporting design, so we can discuss it separately from the rest

msau42 commented:
I removed the capacity api section, and will look at addressing it separately.

So the design proposal here includes:
* Basic changes to scheduler and PV controller to delay dynamic provisioning and pass the selected node to the provisioner
* API changes to allow an administrator to restrict topologies in the storageclass (replacement for plugin-specific zone/zones parameters today)

@saad-ali @thockin @jsafrane PTAL

verult commented on contributors/design-proposals/storage/volume-topology-scheduling.md:
````diff
@@ -266,41 +304,565 @@ const (
+    // Each entry in the map is ANDed.
+    MatchNodeLabels map[string]string
+}
+```
````
> ```
> StorageClass:
>  topologyConstraints:
>    - zone in [a, b]
>      rack in ["1", "2"]
> ```
Wouldn't this allow "zone=b,rack=1"? To entirely capture the original topology, you'd need something like:
```
StorageClass:
  topologyConstraints:
    - zone in [a]
      rack in ["1"]
    - zone in [b]
      rack in ["2"]
```

One inconvenience with the selector approach is there are multiple ways to specify certain kinds of topology. For example, the following 3 are equivalent:
```
StorageClass:
  topologyConstraints:
    - zone in [a]
      rack in ["1"]
    - zone in [b]
      rack in ["1"]
    - zone in [b]
      rack in ["2"]
```
```
StorageClass:
  topologyConstraints:
    - zone in [a,b]
      rack in ["1"]
    - zone in [b]
      rack in ["2"]
```
```
StorageClass:
  topologyConstraints:
    - zone in [a]
      rack in ["1"]
    - zone in [b]
      rack in ["1", "2"]
```

I could potentially see users being confused which form Kubernetes expects, when in reality either of them work. To me it seems better to have one layer of ORs instead of two.

msau42 replied:
My updated example [here](https://github.com/kubernetes/community/pull/1857/files#diff-9d29da618fae1244e58d194d3d70e2adR554) correctly represents (one way) the example topology.  I think having multiple representations is fine as long as they evaluate the same, and it is consistent with how selectors are specified for other Kubernetes objects.

msau42 commented on contributors/design-proposals/storage/volume-topology-scheduling.md:
```diff
@@ -328,30 +890,62 @@ avoid these error conditions are to:
 
 Dynamic provisioning will also be skipped if
-VolumBindingWaitForFirstConsumer is set.  The scheduler will signal to
+VolumBindingWaitForFirstConsumer is set. The scheduler will signal to
```
1. In-tree provisioners will be coordinated by the PV controller.
2. The external-provisioner library in external-storage will be updated to also coordinate this.
3. The CSI external-provisioner will coordinate this.
4. If you don't use the external-provisioner library, then yes, you need to handle this in your own provisioner.

lichuqiang commented on contributors/design-proposals/storage/volume-topology-scheduling.md:
```diff
@@ -225,82 +273,477 @@ Both binding decisions of:
+    // Restrict the node topologies where volumes can be dynamically provisioned.
+    // Each volume plugin defines its own supported topology specifications.
+    // Each entry in AllowedTopologies is ORed.
+    AllowedTopologies *[]TopologySelector
```
Does it have to be a pointer? I think it fairly enough to make it an array, and just leave it empty if we don't wanna specify the field.

and I tried to make it a pointer, but when running `make update`, seems the update scripts still treat the field as array, and tried to range over it:
```
# k8s.io/kubernetes/vendor/k8s.io/api/storage/v1
vendor/k8s.io/api/storage/v1/generated.pb.go:169:8: invalid argument m.AllowedTopologies (type *[]TopologySelector) for len
vendor/k8s.io/api/storage/v1/generated.pb.go:170:17: cannot range over m.AllowedTopologies (type *[]TopologySelector)
vendor/k8s.io/api/storage/v1/generated.pb.go:348:8: invalid argument m.AllowedTopologies (type *[]TopologySelector) for len
vendor/k8s.io/api/storage/v1/generated.pb.go:349:15: cannot range over m.AllowedTopologies (type *[]TopologySelector)
```

So if we have to make it this way, we may need a fix on the update scripts.

msau42 replied:
Oops yes this does not need a pointer. I mistakenly left it from the last iteration when it was a struct

lichuqiang replied:
so `[]*TopologySelector` ?

msau42 replied:
`[] TopologySelector`

saad-ali commented on contributors/design-proposals/storage/volume-topology-scheduling.md:
```diff
@@ -1,24 +1,32 @@
-provisioned PVs.
+* Allow a Pod to request one or more Persistent Volumes (PV) with topology that
+  are compatible with the Pod's other scheduling constraints, such as resource
+  requirements and affinity/anti-affinity policies.
```
nit: This still doesn't sound very clear to me.

Pods don't request anything. Pods specify things that users request. So, to me, this reads "Allow a user to to specify in a pod one or more PV with topology that is compatible with the pod's other constraints."

That still doesn't make sense to me. Do we prevent that now? Is that not possible now?

The following sound a lot more clear to me:
* `Allow the Kubernetes scheduler to influence where a volume is provisioned based on scheduling constraints on the pod referencing the volume, such as pod resource requirements and affinity/anti-affinity policies.`
* `Ensure volume topology is used as a constraint when scheduling a pod referencing one or more Persistent Volumes (PV) with topology.`

These may be implied, but I'd like to be very explicit about them.

saad-ali commented on contributors/design-proposals/storage/volume-topology-scheduling.md:
```diff
@@ -1,24 +1,32 @@
+* Allow a Pod to request one or more Persistent Volumes (PV) with topology that
+  are compatible with the Pod's other scheduling constraints, such as resource
+  requirements and affinity/anti-affinity policies.
+* Support arbitrary PV topology domains (i.e. node, rack, zone, foo, bar).
```
Maybe expand this to explicitly state that we want to this without having to encode as first class objects in the k8s API.

saad-ali commented on contributors/design-proposals/storage/volume-topology-scheduling.md:
```diff
@@ -1,24 +1,32 @@
+  are compatible with the Pod's other scheduling constraints, such as resource
+  requirements and affinity/anti-affinity policies.
+* Support arbitrary PV topology domains (i.e. node, rack, zone, foo, bar).
+* Support topology for statically created PVs and dynamically provisioned PVs.
```
nit: personal preference "pre-provisioned" instead of "statically created". The symmetry with dynamically provisioned is nicer =)

msau42 commented:
This is unicorning too frequently.  I addressed all the comments here, and opened up a new PR https://github.com/kubernetes/community/pull/1857

/close

jheiss commented:
@msau42 Did you mean to refer to #2168 when closing this? (You referred to this same issue in your last comment.)

msau42 commented:
Thanks, updated!
